src/RepoDriver.sol

```diff
@@ -0,0 +1,530 @@
+    }
+
+    struct RepoDriverAnyApiStorage {
+        /// @notice The requested user owner updates.
+        mapping(bytes32 requestId => uint256 userId) requestedUpdates;
+        /// @notice The new address of the AnyApi operator.
+        OperatorInterface operator;
+        /// @notice The fee in Link for each user owner
+        /// update request when the driver is covering the cost.
+        uint96 defaultFee;
+        /// @notice The AnyApi job ID used for requesting user owner updates.
+        bytes32 jobId;
+        /// @notice The AnyApi requests counter used as a nonce when calculating the request ID.
+        uint256 nonce;
+    }
+
+    /// @param _dripsHub The drips hub to use.
+    /// @param forwarder The ERC-2771 forwarder to trust. May be the zero address.
+    /// @param _driverId The driver ID to use when calling DripsHub.
+    constructor(DripsHub _dripsHub, address forwarder, uint32 _driverId)
+        ERC2771Context(forwarder)
+    {
+        dripsHub = _dripsHub;
+        driverId = _driverId;
+        // slither-disable-next-line uninitialized-local
+        address _linkToken;
+        if (block.chainid == CHAIN_ID_ETHEREUM) _linkToken = LINK_TOKEN_ETHEREUM;
+        else if (block.chainid == CHAIN_ID_GOERLI) _linkToken = LINK_TOKEN_GOERLI;
+        else revert("Unsupported chain");
+        linkToken = LinkTokenInterface(_linkToken);
+    }
+
+    modifier onlyOwner(uint256 userId) {
+        require(_msgSender() == ownerOf(userId), "Caller is not the user owner");
+        _;
+    }
+
+    /// @notice Calculates the user ID.
+    /// Every user ID is a 256-bit integer constructed by concatenating:
+    /// `driverId (32 bits) | forgeId (8 bits) | nameEncoded (216 bits)`.
+    /// When `forge` is GitHub and `name` is at most 27 bytes long,
+    /// `forgeId` is 0 and `nameEncoded` is `name` right-padded with zeros
+    /// When `forge` is GitHub and `name` is longer than 27 bytes,
+    /// `forgeId` is 1 and `nameEncoded` is the lower 27 bytes of the hash of `name`.
+    /// When `forge` is GitLab and `name` is at most 27 bytes long,
+    /// `forgeId` is 2 and `nameEncoded` is `name` right-padded with zeros
+    /// When `forge` is GitLab and `name` is longer than 27 bytes,
+    /// `forgeId` is 3 and `nameEncoded` is the lower 27 bytes of the hash of `name`.
+    /// @param forge The forge where the repository is stored.
+    /// @param name The name of the repository.
+    /// For GitHub and GitLab it must follow the `user_name/repository_name` structure
+    /// and it must be formatted identically as in the repository's URL,
+    /// including the case of each letter and special characters being removed.
+    /// @return userId The user ID.
+    function calcUserId(Forge forge, bytes memory name) public view returns (uint256 userId) {
+        uint8 forgeId;
+        uint216 nameEncoded;
+        if (forge == Forge.GitHub) {
+            if (name.length <= 27) {
+                forgeId = 0;
```

Comment:
The enum seems a bit over-engineered, if we still need a `uint8` helper variable here.
```solidity
function calcUserId(uint8 forge, bytes memory name)
```

Would avoid it. We could introduce constants `FORGE_GITHUB`.

Reply:
We can also keep it like that but it resulted in this `helper variable` here.
Maybe you can do it `assembly` and use the enum for it (for generating the hash). It would also simplify these `if` statements.

Reply:
I don't get the idea, could you paste the (pseudo) code? I agree that the `if`s here aren't clever, my goal was to make the code legible. If I wanted to make it clever, I'd probably do a bit shifting and a binary flag for name being long, but that wouldn't be an obvious approach.

Reply:
Fixed it. Tests are passing.

Reply:
Ah, I misunderstood the function.

Reply:
The problem is that `shr(8, forge)` always returns `0`, `forge` is an 8-bit number. I'm quite certain that this code doesn't need assembly though :D

Reply:
This would only work if the enum contains explicit all 4 forgeIds.

```solidity
enum Forge {
    GitHub,
    GithubLong,
    GitLab
    GitLabLong
}
```

Reply:
> The problem is that shr(8, forge) always returns 0, forge is an 8-bit number. I'm quite certain that this code doesn't need assembly though :D
Definitely.  Just playing with ideas.